File: engines/python/agentce/assess.py
"""Assemble assertions by evaluating a catalog against each subject (SPEC §7, §9).

For every subject the engine builds that subject's graph, then for every control in every applied
catalog it decides the outcome: ``not_applicable`` when the control's role does not match or the shape
targets nothing, ``insufficient_evidence`` when the control's minimum evidence (event type and source
class) is absent, and otherwise the structural verdict against the control's tolerance. Supporting
verdicts carry evidence pointers to the focus nodes so DC-5 holds. Only rung-2 structural controls are
evaluated in Phase 1; other rungs become ``not_assessed`` until their evaluators land.
"""

from __future__ import annotations

from typing import Any

from .applicability import effective_roles
from .assertions import Assertion, EvidencePointer
from .canonical import sha256_hex
from .catalog import Catalog, ControlSpec
from .domain import DomainBinding
from .graph import build_graph
from .iri import event_iri
from .profile import Profile, Subject
from .store import GraphStore
from .structural import evaluate_shape, within_tolerance

_EVIDENCE_CAP = 20


def _role_applies(roles: set[str], applies_to_roles: list[str]) -> bool:
    targets = set(applies_to_roles)
    if "both" in targets:
        targets |= {"deployer", "provider"}
    return bool(roles & targets)


def _class_ok(observed: str, required: str) -> bool:
    return required in ("any", "self_report") or observed == required


def requirement_met(events: list[dict[str, Any]], requirement: dict[str, str]) -> bool:
    """Whether ``events`` carries at least one event satisfying one ``minimum_evidence`` entry --
    factored out of :func:`_has_minimum_evidence` so the remediation renderer (SPEC §7) can report,
    per requirement, which of a control's minimum-evidence entries this subject's own events did and
    did not satisfy, using the exact same rule the assessment itself used to reach its outcome."""
    event_type = requirement.get("event")
    required_class = requirement.get("class", "any")
    return any(
        _event_type(e) == event_type
        and _class_ok(str(e.get("agentcesourceclass", "")), required_class)
        for e in events
    )


def _has_minimum_evidence(
    events: list[dict[str, Any]], minimum: list[dict[str, str]]
) -> bool:
    return all(requirement_met(events, requirement) for requirement in minimum)


def _event_type(event: dict[str, Any]) -> str:
    data = event.get("data")
    return (
        str(data["@type"])
        if isinstance(data, dict) and isinstance(data.get("@type"), str)
        else ""
    )


def _window(profile: Profile, events: list[dict[str, Any]]) -> tuple[str, str]:
    window = profile.observation_window
    if "start" in window and "end" in window:
        return window["start"], window["end"]
    times = sorted(str(e.get("time", "")) for e in events if e.get("time"))
    if times:
        return times[0], times[-1]
    return "1970-01-01T00:00:00Z", "1970-01-01T00:00:00Z"


def _evidence(
    events_by_iri: dict[str, dict[str, Any]], focus_nodes: list[str]
) -> list[EvidencePointer]:
    pointers: list[EvidencePointer] = []
    for focus in focus_nodes[:_EVIDENCE_CAP]:
        event = events_by_iri.get(focus)
        if event is None:
            continue
        pointers.append(
            EvidencePointer(
                ref=focus,
                digest="sha256:" + sha256_hex(event),
                source_class=str(event.get("agentcesourceclass") or "self_report"),
            )
        )
    return pointers


def _family(control_id: str) -> str:
    """The control's family: the letter prefix of its own id (SPEC §7.1; ``control.schema.json``
    documents the id pattern ``^[A-Z]{2,4}-[0-9]{2}$`` as "family prefix and two-digit number")."""
    return control_id.split("-", 1)[0] if "-" in control_id else control_id


def _crosswalk(control: ControlSpec) -> list[dict[str, Any]]:
    """Carry each control's own crosswalk entries into every assertion built for it (SPEC §7.3).

    ``verified`` is read from the control's ``verified_against_text`` and coerced to a real
    ``bool`` -- never invented, and never set ``True`` by the engine itself; only a human with
    access to the licensed standard text may flip that flag in the catalog source (human action H4).
    """
    return [
        {
            "framework": entry.get("framework"),
            "clause": entry.get("clause"),
            "verified": bool(entry.get("verified_against_text", False)),
        }
        for entry in control.raw.get("crosswalk", []) or []
    ]


def _assert_control(
    store: GraphStore,
    catalog: Catalog,
    control: ControlSpec,
    subject: Subject,
    roles: set[str],
    events: list[dict[str, Any]],
    events_by_iri: dict[str, dict[str, Any]],
    window: tuple[str, str],
) -> Assertion:
    base: dict[str, Any] = {
        "control": control.id,
        "control_version": control.version,
        "subject": subject.id,
        "rung": control.rung,
        "mode": control.mode,
        "window": window,
        "severity": control.severity,
        "family": _family(control.id),
        "crosswalk": _crosswalk(control),
    }
    if not _role_applies(roles, control.applies_to_roles):
        return Assertion(outcome="not_applicable", population=(0, 0), **base)

    shape = catalog.shape_for(control)
    if control.rung != 2 or shape is None:
        return Assertion(outcome="not_assessed", population=(0, 0), **base)

    applicable, failing, violations = evaluate_shape(
        store, shape, catalog.shapes, control.id
    )
    if not applicable:
        return Assertion(outcome="not_applicable", population=(0, 0), **base)
    if not _has_minimum_evidence(events, control.minimum_evidence):
        return Assertion(
            outcome="insufficient_evidence", population=(len(applicable), 0), **base
        )

    conformant = within_tolerance(len(applicable), len(failing), control.tolerance)
    outcome = "conformant" if conformant else "non-conformant"
    focus_for_evidence = sorted(failing) if failing else applicable
    return Assertion(
        outcome=outcome,
        population=(len(applicable), len(failing)),
        violations=[v.to_json() for v in violations],
        evidence=_evidence(events_by_iri, focus_for_evidence),
        source_class_satisfied=True,
        evidence_strength=control.raw.get("evidence_strength"),
        **base,
    )


#: The outcomes that mean a control was actually judged. A run with none of them evaluated nothing:
#: every pair was inapplicable or beyond the engine's rungs, so its clean tally says nothing.
VERDICT_OUTCOMES = frozenset({"conformant", "non-conformant", "insufficient_evidence"})


def evaluated_nothing(assertions: list[Assertion]) -> bool:
    """True when no assertion reached a verdict (see :data:`VERDICT_OUTCOMES`)."""
    return not any(a.outcome in VERDICT_OUTCOMES for a in assertions)


def index_by_subject(
    accepted: list[dict[str, Any]],
) -> dict[str, list[dict[str, Any]]]:
    """Group ``accepted`` by subject id in one pass over the list (also used by the remediation
    renderer, SPEC §7, to recover the same per-subject event list an assessment run judged)."""
    index: dict[str, list[dict[str, Any]]] = {}
    for event in accepted:
        index.setdefault(str(event.get("subject", "")), []).append(event)
    return index


def assess_subjects(
    accepted: list[dict[str, Any]],
    profile: Profile,
    catalogs: list[Catalog],
    domain: DomainBinding,
) -> list[Assertion]:
    """Evaluate every catalog control against every subject and return the assertions."""
    assertions: list[Assertion] = []
    events_by_subject = index_by_subject(accepted)
    for subject in profile.subjects:
        subject_events = events_by_subject.get(subject.id, [])
        store = build_graph(subject_events, domain=domain)
        events_by_iri = {
            event_iri(str(e["id"])): e for e in subject_events if "id" in e
        }
        roles = set(effective_roles(subject.role))
        window = _window(profile, subject_events)
        for catalog in catalogs:
            for control in catalog.controls:
                assertions.append(
                    _assert_control(
                        store,
                        catalog,
                        control,
                        subject,
                        roles,
                        subject_events,
                        events_by_iri,
                        window,
                    )
                )
    return assertions

File: engines/typescript/src/assess.ts
/**
 * Assemble assertions by evaluating a catalog against each subject (SPEC §7, §9).
 *
 * For every subject the engine builds that subject's graph, then for every control in every applied
 * catalog it decides the outcome: `not_applicable` when the control's role does not match or the shape
 * targets nothing, `insufficient_evidence` when the control's minimum evidence (event type and source
 * class) is absent, and otherwise the structural verdict against the control's tolerance. Supporting
 * verdicts carry evidence pointers to the focus nodes so DC-5 holds. Only rung-2 structural controls
 * are evaluated in Phase 1; other rungs become `not_assessed`. This is a faithful port of the reference.
 */

import { effectiveRoles } from "./applicability";
import { type Assertion, type EvidencePointer, makeAssertion } from "./assertions";
import { sha256Hex } from "./canonical";
import { type Catalog, type ControlSpec, shapeFor } from "./catalog";
import type { DomainBinding } from "./domain";
import { buildGraph } from "./graph";
import { eventIri } from "./iri";
import type { Profile, Subject } from "./profile";
import type { GraphStore } from "./store";
import { evaluateShape, violationToJson, withinTolerance } from "./structural";
import { byteCompare } from "./util";

type Event = Record<string, unknown>;

const EVIDENCE_CAP = 20;
const ROLES = ["deployer", "provider"];

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function roleApplies(roles: Set<string>, appliesToRoles: string[]): boolean {
  const targets = new Set(appliesToRoles);
  if (targets.has("both")) {
    for (const r of ROLES) {
      targets.add(r);
    }
  }
  for (const r of roles) {
    if (targets.has(r)) {
      return true;
    }
  }
  return false;
}

function classOk(observed: string, required: string): boolean {
  return required === "any" || required === "self_report" || observed === required;
}

function eventType(event: Event): string {
  const data = event.data;
  return isRecord(data) && typeof data["@type"] === "string" ? data["@type"] : "";
}

function hasMinimumEvidence(events: Event[], minimum: Array<Record<string, string>>): boolean {
  for (const requirement of minimum) {
    const wantType = requirement.event;
    const requiredClass = requirement.class ?? "any";
    if (
      !events.some(
        (e) =>
          eventType(e) === wantType && classOk(String(e.agentcesourceclass ?? ""), requiredClass),
      )
    ) {
      return false;
    }
  }
  return true;
}

function window(profile: Profile, events: Event[]): [string, string] {
  const w = profile.observationWindow;
  if ("start" in w && "end" in w) {
    return [w.start as string, w.end as string];
  }
  const times = events
    .map((e) => String(e.time ?? ""))
    .filter((t) => t)
    .sort(byteCompare);
  if (times.length > 0) {
    return [times[0] as string, times[times.length - 1] as string];
  }
  return ["1970-01-01T00:00:00Z", "1970-01-01T00:00:00Z"];
}

function evidence(eventsByIri: Map<string, Event>, focusNodes: string[]): EvidencePointer[] {
  const pointers: EvidencePointer[] = [];
  for (const focus of focusNodes.slice(0, EVIDENCE_CAP)) {
    const event = eventsByIri.get(focus);
    if (event === undefined) {
      continue;
    }
    pointers.push({
      ref: focus,
      digest: `sha256:${sha256Hex(event)}`,
      sourceClass: String(event.agentcesourceclass || "self_report"),
    });
  }
  return pointers;
}

function assertControl(
  store: GraphStore,
  catalog: Catalog,
  control: ControlSpec,
  subject: Subject,
  roles: Set<string>,
  events: Event[],
  eventsByIri: Map<string, Event>,
  win: [string, string],
): Assertion {
  const base = {
    control: control.id,
    controlVersion: control.version,
    subject: subject.id,
    rung: control.rung,
    mode: control.mode,
    window: win,
  };
  if (!roleApplies(roles, control.appliesToRoles)) {
    return makeAssertion({ ...base, outcome: "not_applicable", population: [0, 0] });
  }

  const shape = shapeFor(catalog, control);
  if (control.rung !== 2 || shape === null) {
    return makeAssertion({ ...base, outcome: "not_assessed", population: [0, 0] });
  }

  const [applicable, failing, violations] = evaluateShape(store, shape, catalog.shapes, control.id);
  if (applicable.length === 0) {
    return makeAssertion({ ...base, outcome: "not_applicable", population: [0, 0] });
  }
  if (!hasMinimumEvidence(events, control.minimumEvidence)) {
    return makeAssertion({
      ...base,
      outcome: "insufficient_evidence",
      population: [applicable.length, 0],
    });
  }

  const conformant = withinTolerance(applicable.length, failing.size, control.tolerance);
  const focusForEvidence = failing.size > 0 ? [...failing].sort(byteCompare) : applicable;
  return makeAssertion({
    ...base,
    outcome: conformant ? "conformant" : "non-conformant",
    population: [applicable.length, failing.size],
    violations: violations.map(violationToJson),
    evidence: evidence(eventsByIri, focusForEvidence),
    sourceClassSatisfied: true,
    evidenceStrength:
      typeof control.raw.evidence_strength === "string" ? control.raw.evidence_strength : null,
  });
}

/** Group `accepted` by subject id in one pass over the list. */
function indexBySubject(accepted: Event[]): Map<string, Event[]> {
  const index = new Map<string, Event[]>();
  for (const event of accepted) {
    const key = String(event.subject ?? "");
    const events = index.get(key);
    if (events) {
      events.push(event);
    } else {
      index.set(key, [event]);
    }
  }
  return index;
}

/** Evaluate every catalog control against every subject and return the assertions. */
export function assessSubjects(
  accepted: Event[],
  profile: Profile,
  catalogs: Catalog[],
  domain: DomainBinding,
): Assertion[] {
  const assertions: Assertion[] = [];
  const eventsBySubject = indexBySubject(accepted);
  for (const subject of profile.subjects) {
    const subjectEvents = eventsBySubject.get(subject.id) ?? [];
    const store = buildGraph(subjectEvents, { domain });
    const eventsByIri = new Map<string, Event>();
    for (const event of subjectEvents) {
      if ("id" in event) {
        eventsByIri.set(eventIri(String(event.id)), event);
      }
    }
    const roles = new Set(effectiveRoles(subject.role));
    const win = window(profile, subjectEvents);
    for (const catalog of catalogs) {
      for (const control of catalog.controls) {
        assertions.push(
          assertControl(store, catalog, control, subject, roles, subjectEvents, eventsByIri, win),
        );
      }
    }
  }
  return assertions;
}

/** Outcomes that count as a verdict was reached at all (a run that assessed nothing reaches none). */
const VERDICT_OUTCOMES = new Set(["conformant", "non-conformant", "insufficient_evidence"]);

/** True when no assertion reached a verdict (see {@link VERDICT_OUTCOMES}). */
export function evaluatedNothing(assertions: Assertion[]): boolean {
  return !assertions.some((a) => VERDICT_OUTCOMES.has(a.outcome));
}
